src/verilog/verilog_typecheck_expr.cpp

```diff
@@ -1614,6 +1628,34 @@ void verilog_typecheck_exprt::implicit_typecast(
       return;
     }
   }
+  else if(src_type.id() == ID_verilog_assignment_pattern)
+  {
+    DATA_INVARIANT(
+      expr.id() == ID_verilog_assignment_pattern,
+      "verilog_assignment_pattern expression expected");
+    if(dest_type.id() == ID_struct)
+    {
+      auto &struct_type = to_struct_type(dest_type);
+      auto &components = struct_type.components();
+
+      if(expr.operands().size() != components.size())
+      {
+        throw errort().with_location(expr.source_location())
+          << "number of expressions does not match number of struct members";
+      }
+
+      for(std::size_t i = 0; i < components.size(); i++)
+      {
+        // rec. call
+        implicit_typecast(expr.operands()[i], components[i].type());
+      }
+
+      // turn into struct expression
+      expr.id(ID_struct);
+      expr.type() = dest_type;
+      return;
+    }
```

Comment:
The earlier comment said these could become structs or arrays -- are the latter just not yet implemented?

Reply:
Array case added.